content/blog/2019-10-3-tdd-gimmick/index.md

```diff
@@ -0,0 +1,196 @@
+- once the error is spotted, it becomes easier to rewrite the bad test or to add tests for missing edge cases.
+
+Since refactoring is easier, this whole process happens much more quickly. And yes, while fixing one issue you're less likely to introduce a a regression even in this new feature.
+
+## This is all horribly complicated! I'm a pretty good programmer, I don't need all this
+
+I felt the same way when first starting out. Even now, when discussing some of the finer points of TDD it can feel a little overly-cautious and pessimistic. The thing is though, if there's more than 2 programmers working on a project it's just not possible to always be available for everyone.
+
+Imagine you've written a function, and someone wants to make a change to it. If they knew that the tests were high-quality, they could just figure it out themselves and be fairly confident that they weren't doing anything wrong. In the absence of a test-suite they would (hopefully) come and sit on your head and work things out with you.
+
+In the worst case, they're under a lot of pressure so they make the change that re-introduces some subtle bug that you'd already taken care of initially. The bug might only get discovered months later, at which point maybe the dev who introduced the problematic-change is on leave and the commit message just says "fixed issue" so now you're really screwed.
+
+At it's best, TDD is a way to communicate your decisions without needing to write extensive documentation that no one will ever read (or maintain when they make changes). Good-quality unit tests save such an incredible amount of time, and ultimately that's more time that you have to work in peace.
+
+## Are there any cases where one shouldn't use TDD?
+
+For application development, almost everything is a good candidate for TDD. The one exception for frontend development would be if your function returns a particular DOM structure, there's no point writing a test like `Assert <div>Hello World!</div>` and then having a function that just returns `<div>Hello World!</div>`.
+
+You _could_ do it, but then tomorrow if you need to return `<div>Hello <b>World!</b></div>` should the test really break? I mean, I added an enhancement, the old test should "technically" still be valid.
+
+This doesn't mean that all view-related things cannot be done via TDD though. You might have a rock-solid reducer implemented, but if you forget to actually fire an action when a button is clicked then your app is still buggy. Packages like [react-testing-library](https://github.com/testing-library/react-testing-library) help with that problem.
+
+## Any others?
+
+Yes, if you're writing a new package, and there's two potential APIs that would both work equally fine but they both have different (say) performance implications based on their implementation then that might be a bad use-case for TDD.
```

Comment:
Shouldn't be this a perfect case for TDD? The tests should be the same, the implementation can vary.

Reply:
made a change, let me know if this is clearer.